lib/Optimizer/Dialect/Quake/QuakeOps.cpp

```diff
@@ -83,6 +89,49 @@ void quake::AllocaOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
   patterns.add<FuseConstantToAllocaPattern>(context);
 }
 
+//===----------------------------------------------------------------------===//
+// Concat
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct ConcatSizePattern : public OpRewritePattern<quake::ConcatOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::ConcatOp concat,
+                                PatternRewriter &rewriter) const override {
+    if (concat.getType().hasSpecifiedSize())
+      return failure();
+
+    // Walk the arguments and sum them, if possible.
+    std::size_t sum = 0;
+    for (auto opnd : concat.getQbits()) {
+      if (auto veqTy = dyn_cast<quake::VeqType>(opnd.getType())) {
+        if (!veqTy.hasSpecifiedSize())
+          return failure();
+        sum += veqTy.getSize();
+      }
+      assert(isa<quake::RefType>(opnd.getType()));
```

Comment:
Are we missing a `continue`? I read this and see a check that the type is a `veq`, then we increment `sum`, then we fall out of the if and are trying to assert that the type is a `ref`.